### PR TITLE
Prevent email generation failure

### PR DESCRIPTION
**Pull request by shtukas**
## What does this change?

We have problem by which, in some circumstances, emails are generated without their embedded styles. When this happens, an endpoint like this one

https://www.theguardian.com/politics/2018/jul/16/theresa-may-ready-cave-in-hardline-brexiters-customs-demands/email

Present an HTML document without the embedded styles

screenshot 2018-12-14 at 11 51 29-cut

Investigation revealed that the problem was caused by a CSS Parser that is not thread safe. 

When that library was first used, we knew there was a problem ( https://github.com/guardian/frontend/pull/15720 ), but it was not handled the best way and a `Retry` was used to mitigate it. Today's refactoring prevent the same object from being reused between requests.

It is not impossible that a follow up PR will come to get rid of the Retry all together (more investigation is required to make sure it is not needed). 

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

**Comment by shtukas** [comment]
... and thanks @SiAdcock and @joelochlann for help :)

File: common/app/common/InlineStyles.scala [hunk 1]
```diff
@@ -93,27 +93,26 @@ object InlineStyles {
...
-        case Failure(_) =>
-          (inline, head :+ element.html)
-        case Success(sheet) =>
-          val (styles, others) = seq(sheet.getCssRules).partition(isStyleRule)
-          val (inlineStyles, headStyles) = styles.flatMap(CSSRule.fromW3).flatten.partition(_.canInline)
-          val newHead = (headStyles.map(_.toString) ++ others.map(_.getCssText)).mkString("\n")
-
-          (inline ++ inlineStyles, (head :+ newHead).filter(_.nonEmpty))
+      synchronized {
+        Retry(3)(cssParser.parseStyleSheet(source, null, null)) { (exception, attemptNumber) =>
```
Comment (mchv) [inline common/app/common/InlineStyles.scala]: The code becomnes:
```scala
val cssParser = new CSSOMParser(new SACParserCSS3())
cssParser.parseStyleSheet
...
```

File: common/app/common/InlineStyles.scala [hunk 2]
```diff
@@ -93,27 +93,26 @@ object InlineStyles {
...
-      } match {
-        case Failure(_) =>
-          (inline, head :+ element.html)
-        case Success(sheet) =>
-          val (styles, others) = seq(sheet.getCssRules).partition(isStyleRule)
-          val (inlineStyles, headStyles) = styles.flatMap(CSSRule.fromW3).flatten.partition(_.canInline)
-          val newHead = (headStyles.map(_.toString) ++ others.map(_.getCssText)).mkString("\n")
-
-          (inline ++ inlineStyles, (head :+ newHead).filter(_.nonEmpty))
+      synchronized {
```
[the remark this answers is not in the document]
Comment (mchv) [reply common/app/common/InlineStyles.scala]: i.e you only need the `synchronized ` abstraction if you try to reuse the same instance across multiple method invocations. If your method is `pure` (do not depend on a global mutable object), then you are fine

[the remark this answers is not in the document]
Comment (AWare) [reply common/app/common/InlineStyles.scala]: We discussed this and it wasn't obvious whether creating a new parser per invocation or just sticking the invocation in a `synchronized` was the best call here.

**Comment by shtukas** [comment]
@mchv Thanks for the above comments, I am replying to all of them at once here. 

We did try getting rid of the parser singleton (the one instantiated by the `InlineStyles` object) and instead create a new one per request, as you suggested. ( Initially I has simply written it
```
Retry(3)((new CSSOMParser(new SACParserCSS3())).parseStyleSheet(source, null, null)) 
```
 ), and truth to be told it solved the problem (I could not longer reproduce the issue), but we were not user that we would not then run into memory problems in case the object is not correctly cleaned up.

That having been said, you are right. Update coming :)

**Review by AWare — APPROVED** [review]

**Comment by mchv** [comment]
> were not sure that we would not then run into memory problems in case the object is not correctly cleaned up.

Sharing a bit of knowledge. The garbage collector will detect that the object lifetime is short, only for the method, and therefore will keep the object in a special section of the heap called `Eden` (you can read more about this on this [excellent article](https://plumbr.io/handbook/garbage-collection-in-java#eden), it will never move to the `Survivor` spaces.

What you are fearing is a `high allocation rate` which is an important concern but that you should in general completely avoid , as garbage collectors are optimised (especially generational like `G1`) for short lived objects, and you should not tried to optimised unless you have measured and tweak your garbage collector options.

Further more, immutable data structures, like Scala `List` for instance, will have an incredibly high allocation rate, and you are likely to see them as a memory problem before other objects. Obviously there could always exist an object (Database connection for instance) which have high performance initialisation cost or memory consumption, that you need to cache and avoid recreating on every method invocation.

If you are interested to read more, you should read [this article](https://mechanical-sympathy.blogspot.com/2013/07/java-garbage-collection-distilled.html)

**Pull request merged**